driver-core/src/test/java/com/datastax/driver/core/CCMTestsSupport.java

```diff
@@ -0,0 +1,687 @@
+            this.ccmBridge = ccmBridge;
+            this.nodes = nodes;
+            this.erroredOut = false;
+            Runtime.getRuntime().addShutdownHook(new Thread() {
+                @Override
+                public void run() {
+                    CCMTestContext.this.stop();
+                }
+            });
+        }
+
+        private void start() {
+            LOGGER.debug("Starting " + this);
+            for (CCMNode node : nodes) {
+                ccmBridge.bootstrapNodeWithPorts(
+                        node.n,
+                        node.thriftPort,
+                        node.storagePort,
+                        node.binaryPort,
+                        node.jmxPort,
+                        node.remoteDebugPort);
+            }
+            LOGGER.debug("Started " + this);
+        }
+
+        private void stop() {
+            if (discarded)
+                return;
+            LOGGER.debug("Stopping " + this);
+            if (erroredOut) {
+                executeNoFail(new Runnable() {
+                    @Override
+                    public void run() {
+                        ccmBridge.stop();
+                    }
+                });
+                LOGGER.debug("Error during tests, kept C* logs in " + ccmBridge.getCcmDir());
+            } else {
+                executeNoFail(new Runnable() {
+                    @Override
+                    public void run() {
+                        ccmBridge.remove();
+                    }
+                });
+                executeNoFail(new Runnable() {
+                    @Override
+                    public void run() {
+                        Files.delete(ccmBridge.getCcmDir());
+                    }
+                });
+            }
+            discarded = true;
+            LOGGER.debug("Stopped " + this);
+        }
+
+        @Override
+        public String toString() {
+            return Objects.toStringHelper(this)
+                    .add("id", id)
+                    .add("nodes", nodes)
+                    .add("ccm", ccmBridge)
+                    .toString();
+        }
+    }
+
+    private static class CCMTestContextLoader extends CacheLoader<CCMTestContextKey, CCMTestContext> {
+
+        @Override
+        public CCMTestContext load(CCMTestContextKey key) {
+            return key.createAndStartTestContext();
+        }
+
+    }
+
+    private static class CCMTestContextRemovalListener implements RemovalListener<CCMTestContextKey, CCMTestContext> {
+
+        @Override
+        public void onRemoval(RemovalNotification<CCMTestContextKey, CCMTestContext> notification) {
+            if (notification.getValue() != null) {
+                notification.getValue().stop();
+            }
+        }
+
+    }
+
+    /**
+     * A LoadingCache that stores running CCM clusters.
+     */
+    private static final LoadingCache<CCMTestContextKey, CCMTestContext> CACHE = CacheBuilder.newBuilder()
+            .initialCapacity(5)
```

Comment:
I like this idea but I think we should reduce the maximum size to something smaller (somewhere between 2-5?).  How many instances we'd want to run probably depends on the system it is on, maybe we should set the maximum size to # of processors or something like that?